diff --git a/makefiles/common.mk b/makefiles/common.mk
index fcbf8bc..139ee20 100644
--- a/makefiles/common.mk
+++ b/makefiles/common.mk
@@ -280,4 +280,6 @@ ifeq ($(THEOS_CURRENT_INSTANCE),)
 endif
 THEOS_PACKAGE_VERSION = $(call __simplify,THEOS_PACKAGE_VERSION,$(THEOS_PACKAGE_BASE_VERSION)$(warning THEOS_PACKAGE_VERSION is deprecated. Please migrate to THEOS_PACKAGE_BASE_VERSION.))
 
+THEOS_LINKAGE_TYPE ?= dynamic
+
 $(eval $(call __mod,common.mk))
diff --git a/makefiles/instance/library.mk b/makefiles/instance/library.mk
index 78e6b91..78083f7 100644
--- a/makefiles/instance/library.mk
+++ b/makefiles/instance/library.mk
@@ -14,11 +14,24 @@ ifeq ($(_LOCAL_LIBRARY_EXTENSION),-)
 	_LOCAL_LIBRARY_EXTENSION =
 endif
 
+_LOCAL_ARCHIVE_EXTENSION = $(or $($(THEOS_CURRENT_INSTANCE)_ARCHIVE_EXTENSION),$(TARGET_AR_EXT))
+ifeq ($(_LOCAL_ARCHIVE_EXTENSION),-)
+	_LOCAL_ARCHIVE_EXTENSION = 
+endif
+
+_LOCAL_LINKAGE_TYPE = $(or $($(THEOS_CURRENT_INSTANCE)_LINKAGE_TYPE),$(THEOS_LINKAGE_TYPE))
+
 _THEOS_INTERNAL_LDFLAGS += $(call TARGET_LDFLAGS_DYNAMICLIB,$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION))
 _THEOS_INTERNAL_CFLAGS += $(TARGET_CFLAGS_DYNAMICLIB)
 
 ifeq ($(_THEOS_MAKE_PARALLEL_BUILDING), no)
+ifeq ($(_LOCAL_LINKAGE_TYPE),static)
+internal-library-all_:: $(_OBJ_DIR_STAMPS) $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_ARCHIVE_EXTENSION)
+else ifeq ($(_LOCAL_LINKAGE_TYPE),both)
+internal-library-all_:: $(_OBJ_DIR_STAMPS) $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION) $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_ARCHIVE_EXTENSION)
+else
 internal-library-all_:: $(_OBJ_DIR_STAMPS) $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION)
+endif
 else
 internal-library-all_:: $(_OBJ_DIR_STAMPS)
 	$(ECHO_MAKE)$(MAKE) -f $(_THEOS_PROJECT_MAKEFILE_NAME) $(_THEOS_MAKEFLAGS) \
@@ -26,15 +39,27 @@ internal-library-all_:: $(_OBJ_DIR_STAMPS)
 		_THEOS_CURRENT_TYPE=$(_THEOS_CURRENT_TYPE) THEOS_CURRENT_INSTANCE=$(THEOS_CURRENT_INSTANCE) _THEOS_CURRENT_OPERATION=compile \
 		THEOS_BUILD_DIR="$(THEOS_BUILD_DIR)" _THEOS_MAKE_PARALLEL=yes
 
+ifeq ($(_LOCAL_LINKAGE_TYPE),static)
+internal-library-compile: $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_ARCHIVE_EXTENSION)
+else ifeq ($(_LOCAL_LINKAGE_TYPE),both)
+internal-library-compile: $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION) $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_ARCHIVE_EXTENSION)
+else
 internal-library-compile: $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION)
 endif
+endif
 
 $(eval $(call _THEOS_TEMPLATE_DEFAULT_LINKING_RULE,$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION)))
+$(eval $(call _THEOS_TEMPLATE_ARCHIVE_LINKING_RULE,$(THEOS_CURRENT_INSTANCE)$(_LOCAL_ARCHIVE_EXTENSION)))
 
 ifneq ($($(THEOS_CURRENT_INSTANCE)_INSTALL),0)
 internal-library-stage_::
 	$(ECHO_NOTHING)mkdir -p "$(THEOS_STAGING_DIR)$(LOCAL_INSTALL_PATH)/"$(ECHO_END)
+ifneq (static,$(_LOCAL_LINKAGE_TYPE))
 	$(ECHO_NOTHING)cp $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_LIBRARY_EXTENSION) "$(THEOS_STAGING_DIR)$(LOCAL_INSTALL_PATH)/"$(ECHO_END)
 endif
+ifneq (,$(filter static both,$(_LOCAL_LINKAGE_TYPE)))
+	$(ECHO_NOTHING)cp $(THEOS_OBJ_DIR)/$(THEOS_CURRENT_INSTANCE)$(_LOCAL_ARCHIVE_EXTENSION) "$(THEOS_STAGING_DIR)$(LOCAL_INSTALL_PATH)/"$(ECHO_END)
+endif
+endif
 
 $(eval $(call __mod,instance/library.mk))
diff --git a/makefiles/instance/rules.mk b/makefiles/instance/rules.mk
index 62287a8..62e2106 100644
--- a/makefiles/instance/rules.mk
+++ b/makefiles/instance/rules.mk
@@ -362,4 +362,67 @@ endif
 
 endef
 
+define _THEOS_TEMPLATE_ARCHIVE_LINKING_RULE
+ifeq ($(TARGET_LIPO),)
+$$(THEOS_OBJ_DIR)/$(1): $$(OBJ_FILES_TO_LINK)
+ifneq ($(2),nowarn)
+ifeq ($$(OBJ_FILES_TO_LINK),)
+	$$(WARNING_EMPTY_LINKING)
+endif
+endif
+ifeq ($$(_THEOS_CURRENT_TYPE),subproject)
+	$$(ECHO_LINKING)$$(ECHO_UNBUFFERED)$$(TARGET_LIBTOOL) -static -o "$$@" $$^$$(ECHO_END)
+else
+	$$(ECHO_LINKING)$$(ECHO_UNBUFFERED)$$(TARGET_LIBTOOL) -static -o "$$@" $$^$$(ECHO_END)
+ifeq ($(SHOULD_STRIP),$(_THEOS_TRUE))
+	$$(ECHO_STRIPPING)$$(ECHO_UNBUFFERED)$$(TARGET_STRIP) $$(ALL_STRIP_FLAGS) "$$@"$$(ECHO_END)
+endif
+endif
+else ifeq ($(THEOS_CURRENT_ARCH),)
+ifeq ($(_THEOS_LIBRARY_TYPE),static)
+
+ARCH_FILES_TO_LINK := $(addsuffix /$(1),$(addprefix $(THEOS_OBJ_DIR)/,$(TARGET_ARCHS)))
+$$(THEOS_OBJ_DIR)/%/$(1): $(__ALL_FILES)
+	@mkdir -p $(THEOS_OBJ_DIR)/$$*
+	$(ECHO_MAKE)$(MAKE) -f $(_THEOS_PROJECT_MAKEFILE_NAME) --no-print-directory --no-keep-going \
+		internal-$(_THEOS_CURRENT_TYPE)-$(_THEOS_CURRENT_OPERATION) \
+		_THEOS_CURRENT_TYPE="$(_THEOS_CURRENT_TYPE)" \
+		THEOS_CURRENT_INSTANCE="$(THEOS_CURRENT_INSTANCE)" \
+		_THEOS_CURRENT_OPERATION="$(_THEOS_CURRENT_OPERATION)" \
+		THEOS_BUILD_DIR="$(THEOS_BUILD_DIR)" \
+		THEOS_CURRENT_ARCH="$$*"
+
+endif
+$(THEOS_OBJ_DIR)/$(1): $$(ARCH_FILES_TO_LINK)
+ifeq ($$(_THEOS_CURRENT_TYPE),subproject)
+	@echo "$$(_THEOS_INTERNAL_LDFLAGS)" > $$(THEOS_OBJ_DIR)/$$(THEOS_CURRENT_INSTANCE).ldflags
+endif
+	$(ECHO_MERGING)$(ECHO_UNBUFFERED)$(TARGET_LIPO) $(foreach ARCH,$(TARGET_ARCHS),-arch $(ARCH) $(THEOS_OBJ_DIR)/$(ARCH)/$(1)) -create -output "$$@"$(ECHO_END)
+
+else
+$$(THEOS_OBJ_DIR)/$(1): $$(OBJ_FILES_TO_LINK)
+ifneq ($(2),nowarn)
+ifeq ($$(OBJ_FILES_TO_LINK),)
+	$$(WARNING_EMPTY_LINKING)
+endif
+endif
+	$$(ECHO_NOTHING)mkdir -p $(shell dirname "$(THEOS_OBJ_DIR)/$(1)")$$(ECHO_END)
+ifeq ($$(_THEOS_CURRENT_TYPE),subproject)
+	$$(ECHO_STATIC_LINKING)$$(ECHO_UNBUFFERED)$$(TARGET_LIBTOOL) -static -o "$$@" $$^$$(ECHO_END)
+	@echo "$$(_THEOS_INTERNAL_LDFLAGS)" > $$(THEOS_OBJ_DIR)/$$(THEOS_CURRENT_INSTANCE).ldflags
+else
+	$$(ECHO_STATIC_LINKING)$$(ECHO_UNBUFFERED)$$(TARGET_LIBTOOL) -static -o "$$@" $$^$$(ECHO_END)
+ifeq ($(SHOULD_STRIP),$(_THEOS_TRUE))
+ifeq ($$(_THEOS_IS_WSL),$(_THEOS_TRUE))
+	mkdir -p "$$(_THEOS_TMP_FOR_WSL)/$$(THEOS_CURRENT_ARCH)"
+	$$(ECHO_STRIPPING)$$(ECHO_UNBUFFERED)cp "$$@" "$$(_THEOS_TMP_FOR_WSL)/$$(THEOS_CURRENT_ARCH)" && $$(TARGET_STRIP) $$(ALL_STRIP_FLAGS) "$$(_THEOS_TMP_FOR_WSL)/$$(THEOS_CURRENT_ARCH)/$(1)" && mv "$$(_THEOS_TMP_FOR_WSL)/$$(THEOS_CURRENT_ARCH)/$(1)" $$(THEOS_OBJ_DIR)$$(ECHO_END)
+else
+	$$(ECHO_STRIPPING)$$(ECHO_UNBUFFERED)$$(TARGET_STRIP) $$(ALL_STRIP_FLAGS) "$$@"$$(ECHO_END)
+endif
+endif
+endif
+endif
+
+endef
+
 $(eval $(call __mod,instance/rules.mk))
diff --git a/makefiles/messages.mk b/makefiles/messages.mk
index c05c191..43790f1 100644
--- a/makefiles/messages.mk
+++ b/makefiles/messages.mk
@@ -61,12 +61,14 @@ ifneq ($(THEOS_CURRENT_ARCH),)
 	ECHO_COMPILING = $(ECHO_BEGIN)$(PRINT_FORMAT_GREEN) "Compiling $< ($(THEOS_CURRENT_ARCH))"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_SWIFTMODULE_HEADER = $(ECHO_BEGIN)$(PRINT_FORMAT_BLUE) "Generating $(notdir $@) ($(THEOS_CURRENT_ARCH))"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_LINKING = $(ECHO_BEGIN)$(PRINT_FORMAT_YELLOW) "Linking $(_THEOS_CURRENT_TYPE) $(THEOS_CURRENT_INSTANCE) ($(THEOS_CURRENT_ARCH))"$(ECHO_END); $(ECHO_PIPEFAIL) (
+	ECHO_STATIC_LINKING = $(ECHO_BEGIN)$(PRINT_FORMAT_YELLOW) "Linking static $(_THEOS_CURRENT_TYPE) $(THEOS_CURRENT_INSTANCE) ($(THEOS_CURRENT_ARCH))"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_LINKING_WITH_STRIP = $(ECHO_BEGIN)$(PRINT_FORMAT_YELLOW) "Linking $(_THEOS_CURRENT_TYPE) $(THEOS_CURRENT_INSTANCE) (with strip, $(THEOS_CURRENT_ARCH))"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_STRIPPING = $(ECHO_BEGIN)$(PRINT_FORMAT_BLUE) "Stripping $(THEOS_CURRENT_INSTANCE) ($(THEOS_CURRENT_ARCH))"$(ECHO_END); $(ECHO_PIPEFAIL) (
 else
 	ECHO_COMPILING = $(ECHO_BEGIN)$(PRINT_FORMAT_GREEN) "Compiling $<"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_SWIFTMODULE_HEADER = $(ECHO_BEGIN)$(PRINT_FORMAT_BLUE) "Generating $(notdir $@)"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_LINKING = $(ECHO_BEGIN)$(PRINT_FORMAT_YELLOW) "Linking $(_THEOS_CURRENT_TYPE) $(THEOS_CURRENT_INSTANCE)"$(ECHO_END); $(ECHO_PIPEFAIL) (
+	ECHO_STATIC_LINKING = $(ECHO_BEGIN)$(PRINT_FORMAT_YELLOW) "Linking static $(_THEOS_CURRENT_TYPE) $(THEOS_CURRENT_INSTANCE)"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_LINKING_WITH_STRIP = $(ECHO_BEGIN)$(PRINT_FORMAT_YELLOW) "Linking $(_THEOS_CURRENT_TYPE) $(THEOS_CURRENT_INSTANCE) (with strip)"$(ECHO_END); $(ECHO_PIPEFAIL) (
 	ECHO_STRIPPING = $(ECHO_BEGIN)$(PRINT_FORMAT_BLUE) "Stripping $(THEOS_CURRENT_INSTANCE)"$(ECHO_END); $(ECHO_PIPEFAIL) (
 endif
diff --git a/makefiles/targets/_common/cygwin.mk b/makefiles/targets/_common/cygwin.mk
index 43a4ef6..e690a2a 100755
--- a/makefiles/targets/_common/cygwin.mk
+++ b/makefiles/targets/_common/cygwin.mk
@@ -1,6 +1,7 @@
 # Variables that are common to all Linux-based targets.
 TARGET_EXE_EXT := .exe
 TARGET_LIB_EXT := .dll
+TARGET_AR_EXT := .a
 
 TARGET_LDFLAGS_DYNAMICLIB = -shared -Wl,-soname,$(1)
 TARGET_CFLAGS_DYNAMICLIB = -fPIC
diff --git a/makefiles/targets/_common/darwin_head.mk b/makefiles/targets/_common/darwin_head.mk
index dc65216..d56044c 100644
--- a/makefiles/targets/_common/darwin_head.mk
+++ b/makefiles/targets/_common/darwin_head.mk
@@ -1,6 +1,7 @@
 # Variables that are common to all Darwin-based targets.
 TARGET_EXE_EXT :=
 TARGET_LIB_EXT := .dylib
+TARGET_AR_EXT := .a
 
 TARGET_LDFLAGS_DYNAMICLIB = -dynamiclib -install_name "$(LOCAL_INSTALL_PATH)/$(1)"
 TARGET_CFLAGS_DYNAMICLIB = 
diff --git a/makefiles/targets/_common/linux.mk b/makefiles/targets/_common/linux.mk
index 3d19dbc..6988cba 100644
--- a/makefiles/targets/_common/linux.mk
+++ b/makefiles/targets/_common/linux.mk
@@ -1,6 +1,7 @@
 # Variables that are common to all Linux-based targets.
 TARGET_EXE_EXT :=
 TARGET_LIB_EXT := .so
+TARGET_AR_EXT := .a
 
 TARGET_LDFLAGS_DYNAMICLIB = -shared -Wl,-soname,$(1)
 TARGET_CFLAGS_DYNAMICLIB = -fPIC
